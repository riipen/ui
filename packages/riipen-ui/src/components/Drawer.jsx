import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

import Backdrop from "./Backdrop";
import ClickAway from "./ClickAway";

class Drawer extends React.Component {
  static propTypes = {
    /**
     * Side from which the drawer will appear.
     */
    anchor: PropTypes.oneOf(["left", "top", "right", "bottom"]),

    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.object,

    /**
     * Callback fired when the component requests to be closed.
     *
     * @param {object} event The event source of the callback.
     */
    onClose: PropTypes.func,

    /**
     * If `true`, the drawer is open.
     */
    open: PropTypes.bool.isRequired
  };

  static defaultProps = {
    anchor: "left",
    open: false
  };

  static contextType = ThemeContext;

  render() {
    const { anchor, children, classes, onClose, open } = this.props;

    if (!open) return null;

    const theme = this.context;

    const className = clsx("drawer", anchor, classes);

    return (
      <React.Fragment>
        <Backdrop />
        <ClickAway onClickAway={onClose}>
          <div className={className}>{children}</div>
        </ClickAway>
        <style jsx>{`
          .drawer {
            background-color: ${theme.palette.common.white};
            display: flex;
            flex-direction: column;
            flex: 1 0 auto;
            height: 100%;
            overflow-y: auto;
            position: fixed;
            top: 0;
            z-index: ${theme.zIndex.middle};
          }

          .left {
            left: 0;
            right: auto;
          }
          .right {
            left: auto;
            right: 0;
          }
          .top {
            top: 0;
            left: 0;
            bottom: auto;
            right: 0;
            height: auto;
            max-height: 100%;
          }
          .bottom {
            top: auto;
            left: 0;
            bottom: 0;
            right: 0;
            height: auto;
            max-height: 100%;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default Drawer;
