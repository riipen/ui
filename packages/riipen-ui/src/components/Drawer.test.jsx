import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Drawer from "./Drawer.jsx";

describe("<Drawer>", () => {
  it("renders without errors", () => {
    let error;
    try {
      mount(<Drawer />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("snapshot test when open is true and onClose is provided", () => {
    const open = true;
    const onClose = jest.fn();

    const wrapper = mount(<Drawer open={open} onClose={onClose} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Drawer().type.defaultProps;

      const wrapper = mount(<Drawer />);

      const component = wrapper.find("Drawer");
      expect(component.props().anchor).toEqual(defaultProps.anchor);
      expect(component.props().open).toEqual(defaultProps.open);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("appends higher order values to default classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-drawer"];

      const wrapper = mount(<Drawer />);

      expect(
        wrapper
          .find("Drawer")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("anchor prop", () => {
    it("sets anchor class name with a valid anchor and open prop is true", () => {
      const anchor = "bottom";
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(
        <Drawer anchor={anchor} open={open} onClose={onClose} />
      );

      expect(
        wrapper
          .find("ClickAway")
          .childAt(0)
          .childAt(0)
          .hasClass(anchor)
      ).toEqual(true);
    });

    it("gives an error when given an invalid anchor", () => {
      const anchor = "leftright";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Drawer anchor={anchor} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("children prop", () => {
    it("passes children to the ClickAway component when open prop is true", () => {
      const child = <h1>The children prop</h1>;
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(
        <Drawer children={child} open={open} onClose={onClose} />
      );

      expect(wrapper.find("ClickAway").contains(child)).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the div inside of ClickAway", () => {
      const classes = ["classOne", "classTwo"];
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(
        <Drawer classes={classes} open={open} onClose={onClose} />
      );

      expect(
        wrapper
          .find("ClickAway")
          .childAt(0)
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
      expect(
        wrapper
          .find("ClickAway")
          .childAt(0)
          .childAt(0)
          .hasClass(classes[1])
      ).toEqual(true);
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Drawer classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("onClose prop", () => {
    it("passes onClose handler to ClickAway", () => {
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(<Drawer open={open} onClose={onClose} />);

      expect(wrapper.find("ClickAway").props().onClickAway).toEqual(onClose);
    });

    it("invokes onClose handler when component requests to be closed", () => {
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(<Drawer open={open} onClose={onClose} />);

      wrapper.find("ClickAway").simulate("click");

      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe("open prop", () => {
    it("renders contents of the component when open is true", () => {
      const open = true;
      const onClose = jest.fn();

      const wrapper = mount(<Drawer open={open} onClose={onClose} />);

      expect(wrapper.find("Drawer").children()).not.toHaveLength(0);
      expect(wrapper.find("Backdrop")).toHaveLength(1);
      expect(wrapper.find("ClickAway")).toHaveLength(1);
    });

    it("does not render contents of the component when open is false", () => {
      const open = false;
      const onClose = jest.fn();

      const wrapper = mount(<Drawer open={open} onClose={onClose} />);

      expect(wrapper.find("Drawer").children()).toHaveLength(0);
    });
  });
});
