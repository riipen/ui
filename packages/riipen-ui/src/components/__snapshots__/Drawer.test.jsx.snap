// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer> default props sets correct default props 1`] = `
<withClasses(Drawer)>
  <Drawer
    anchor="left"
    classes={
      Array [
        "riipen",
        "riipen-drawer",
      ]
    }
    open={false}
  />
</withClasses(Drawer)>
`;

exports[`<Drawer> snapshot test when open is true and onClose is provided 1`] = `
<withClasses(Drawer)
  onClose={[MockFunction]}
  open={true}
>
  <Drawer
    anchor="left"
    classes={
      Array [
        "riipen",
        "riipen-drawer",
      ]
    }
    onClose={[MockFunction]}
    open={true}
  >
    <withClasses(Backdrop)>
      <Backdrop
        classes={
          Array [
            "riipen",
            "riipen-backdrop",
          ]
        }
      >
        <div
          aria-hidden={true}
          className="jsx-1867827232 root riipen riipen-backdrop"
        />
        <JSXStyle
          id="1867827232"
        />
      </Backdrop>
    </withClasses(Backdrop)>
    <withClasses(ClickAway)
      onClickAway={[MockFunction]}
    >
      <ClickAway
        classes={
          Array [
            "riipen",
            "riipen-clickaway",
          ]
        }
        onClickAway={[MockFunction]}
      >
        <div
          className="jsx-3334867826"
          onClick={[Function]}
        >
          <div
            className="jsx-2439981594 drawer left riipen riipen-drawer"
            key=".0"
            onClick={[Function]}
          />
        </div>
        <JSXStyle
          dynamic={
            Array [
              3,
            ]
          }
          id="1169064260"
        />
      </ClickAway>
    </withClasses(ClickAway)>
    <JSXStyle
      dynamic={
        Array [
          "#fff",
          3,
        ]
      }
      id="3412078254"
    />
  </Drawer>
</withClasses(Drawer)>
`;
